packages/preferences-api/src/browser/preference-service.ts

```diff
@@ -0,0 +1,177 @@
+
+    protected readonly readyProviders: PreferenceProvider[] = [];
+
+    protected resolveReady: () => void;
+
+    readonly ready = new Promise<void>(resolve => {
+        this.resolveReady = resolve;
+    });
+
+    protected isReady: boolean = false;
+
+    constructor(protected readonly preferenceProviders: PreferenceProvider[]) {
+        this.preferenceProviders.forEach(preferenceProvider => {
+            preferenceProvider.setClient({
+                onNewPreferences: event => this.onNewPreferences(event)
+            });
+
+            this.toDispose.push(preferenceProvider);
+            preferenceProvider.ready.then(() => {
+                this.readyProviders.push(preferenceProvider);
+                /* Sorted so that providers with more priority i.e workspace > user override user settings if found */
+                this.readyProviders.sort((provider1, provider2) =>
+                    provider1.priority - provider2.priority
+                );
+            });
+        });
+    }
+
+    dispose(): void {
+        this.toDispose.dispose();
+    }
+
+    protected getPreferencesFromProviders() {
+
+        const preferenceChanges: { [preferenceName: string]: PreferenceChange } = {};
+        const deleted = new Set(Object.keys(this.prefCache));
+
+        this.readyProviders.forEach(provider => {
+            const prefs = provider.getPreferences();
+            // tslint:disable-next-line:forin
+            for (const preferenceName in prefs) {
+                deleted.delete(preferenceName);
+                const oldValue = this.prefCache[preferenceName];
+                const newValue = prefs[preferenceName];
+                if (oldValue) {
+                    /* Value changed */
+                    if (!JSONExt.deepEqual(oldValue, newValue)) {
+                        preferenceChanges[preferenceName] = { preferenceName, newValue, oldValue };
+                        this.prefCache[preferenceName] = newValue;
+                    }
+                    /* Value didn't change - Do nothing */
+                } else {
+                    /* New value without old value */
+                    preferenceChanges[preferenceName] = { preferenceName, newValue };
+                    this.prefCache[preferenceName] = newValue;
+                }
+            }
+        });
+
+        /* Deleted values */
+        for (const preferenceName of deleted) {
+            const oldValue = this.prefCache[preferenceName];
+            preferenceChanges[preferenceName] = { preferenceName, oldValue };
+            this.prefCache[preferenceName] = undefined;
+        }
+        // tslint:disable-next-line:forin
+        for (const preferenceName in preferenceChanges) {
+            this.onPreferenceChangedEmitter.fire(preferenceChanges[preferenceName]);
+        }
+    }
+
+    protected onNewPreferences(event: NewPreferencesEvent): void {
+        this.getPreferencesFromProviders();
+    }
+
+    init() {
```

Comment:
Why cannot do it in the constructor?

Reply:
if I do it in the constructor that's where I a dependency loop problem that never ends. Maybe it should be breaked somewhere else than in the preference-service, but I wasn't sure where...